### PR TITLE
annalytics test

### PR DESCRIPTION
**Pull request by tomivm**
is neccesary to test batchGet with the commplete request array? this take more time. test only if the request is sending and responding with a empty array is enought? @sylvansson

### CODE REVIEW

File: test/controllers/analytics.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
+const request = require('supertest');
+const chai = require('chai');
+
+const server = require('../../app');
+const helper = require('../helper');
+
+//Parent block
+describe('analytics API calls', function () {
```
Comment (sylvansson) [inline test/controllers/analytics.js:8]: Since you're testing multiple routes, I would suggest using nested `describe` blocks to group your tests. This makes it easier to read and organize your tests, and to test a single route if needed.

```javascript
describe('Analytics API calls', function() {
  describe('POST /analytics/batchGet', function() {
    it('does something', function() { 
      ... 
    });
  ));

  // We only test this route.
  describe.only('POST /analytics/userActivity/:userId', function() {
    it('does something', function() { 
      ... 
    });
  ));
});
```

I see that the current tests don't use that structure though, so let's see what @martinbedouret thinks.

File: test/controllers/analytics.js [hunk 2]
```diff
@@ -0,0 +1,65 @@
+const request = require('supertest');
```
Comment (sylvansson) [inline test/controllers/analytics.js:1]: This is out of scope for this PR so don't worry about it for now, but it would be nice to rewrite the tests so that we don't need to use the `done` callback.

**Review by sylvansson — COMMENTED** [review]
@tomivm To answer your question, I think it could be useful to [stub](https://sinonjs.org/releases/latest/stubs/) `analyticsreporting.reports.batchGet` during your tests and check that we are passing the right filters, etc. 

@martinbedouret What do you think?

File: test/controllers/analytics.js (hunk 2 above)
Comment (tomivm) [reply test/controllers/analytics.js:1]: can you provide me a tutorial/information link? please ;)

File: test/controllers/analytics.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+  before(async function (done) {
+    this.timeout(5000); //to await the email server process
+    helper.prepareUser(server).then((token, userId) => {
+      authToken = token;
+      userid = userId;
+      done();
+    });
+  });
+
+  it('it should NOT return analytics User data without bearer.', function (done) {
```
[the remark this answers is not in the document]
Comment (sylvansson) [reply test/controllers/analytics.js:21]: That's a good point. Never mind :-)

File: test/controllers/analytics.js (hunk 2 above)
Comment (sylvansson) [reply test/controllers/analytics.js:1]: I don't know a good resource to learn, but you can try this. Make sure you `await` the request, otherwise your test will give you a false positive (it will pass when it shouldn't).

```javascript
it('does something', async function() {
  const res = await request(server)
    .post('/my/route')
    .set('Authorization', `Bearer ${authToken}`)
    .set('Accept', 'application/json')
    .send({})
    .expect('Content-Type', /json/)
    .expect(200);
  expect(res.body.someProperty).to.equal('something');
});
```

(But let's see if @martinbedouret likes this approach first.)

**Comment by tomivm** [comment]
Thanks for your comment! @sylvansson 
i think there are a few processes that we can stub, like batchGet and email server process. I don't know if right now is a good idea to start with that ( i never do that, and I think that we need to use sinon.js ). Maybe is better to finish with these tests like end 2 end (integrated pre Prod test) and then do the tests with stubs. what do you think about this @sylvansson @martinbedouret ?
sory if i dont undertands some concepts and i ask again. thaks for your time!

File: test/controllers/analytics.js (hunk 2 above)
Comment (martinbedouret) [reply test/controllers/analytics.js:1]: @tomivm @sylvansson I like this approach as we work like synchronous test and we forget about the `done`  callbacks.. Good suggestion!!

File: test/controllers/analytics.js (hunk 1 above)
Comment (martinbedouret) [reply test/controllers/analytics.js:8]: agreed!

File: test/controllers/analytics.js (hunk 2 above)
Comment (tomivm) [reply test/controllers/analytics.js:1]: thanks you very much for your explanations! I will study more and start a new PR

Comment (tomivm) [reply test/controllers/analytics.js:1]: thanks very much for yours explanations! I gonna try/study this and start a new PR
